### PR TITLE
Add preprocessors `distance_metrics` and `histogram`

### LINKED ISSUE #2266: New preprocessor: Distance metrics dataset vs. reference
**Is your feature request related to a problem? Please describe.**

To facilitate the comparison of different models against each other or against obersvations, it would be really helpful to have a preprocessor that calculates distance metrics of two datasets. Example for distance metrics can be the RMSE or the Pearson correlation coefficient.

I am thinking of the following signature for the new preprocessor:

```py
def distance_metric(
    products: set[PreprocessorFile] | Iterable[Cube],
    metric: MetricType,
    reference: Optional[Cube] = None,
    coords: Iterable[Coord] | Iterable[str] | None = None,
    keep_reference_dataset: bool = True,
    **kwargs,
) -> set[PreprocessorFile] | CubeList:
```

When used within a recipe, reference datasets should be marked with `reference_for_metric: true`. This signature also allows an easy usage of the preprocessor outside of ESMValTool

**Would you be able to help out?**

Yes!

@axel-lauer @hb326 @LisaBock

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR adds a `distance_metrics` preprocessor that is able to calculate distance metrics between datasets and a reference dataset. In addition, a preprocessor `histogram` is added (which is necessary to calculate one of the metrics). These preprocessors have the following call signatures:

```py
def distance_metric(
    products: set[PreprocessorFile] | Iterable[Cube],
    metric: MetricType,
    reference: Optional[Cube] = None,
    coords: Iterable[Coord] | Iterable[str] | None = None,
    keep_reference_dataset: bool = True,
    **kwargs,
) -> set[PreprocessorFile] | CubeList:
```

```py
def histogram(
    cube: Cube,
    coords: Iterable[Coord] | Iterable[str] | None = None,
    bins: int | Sequence[float] = 10,
    bin_range: tuple[float, float] | None = None,
    weights: np.ndarray | da.Array | bool | None = None,
    normalization: Literal['sum', 'integral'] | None = None,
) -> Cube:
```

If used within a recipe, exactly one dataset in the products which enter `distance_metrics` needs the `reference_for_metric: true` key. Example:

```yaml
datasets:
  - {dataset: BCC-ESM1, project: CMIP6, exp: historical, ensemble: r1i1p1f1, grid: gn}
  - {dataset: bcc-csm1-1, version: v1, project: CMIP5, exp: historical, ensemble: r1i1p1, reference_for_metric: true}

preprocessors:
  calc_rmse:
    regrid:
      target_grid: 3x3
      scheme: linear
    distance_metric:
      metric: rmse
      coords: [latitude, longitude]
      # keep_reference_dataset: false

diagnostics:
  test:
    variables:
      tas:
        mip: Amon
        timerange: '2000/2005'
        preprocessor: calc_rmse
    scripts:
      null
```

Currently supported metrics:
- `rmse`: [root mean square error](https://en.wikipedia.org/wiki/Root-mean-square_deviation)
- `weighted_rmse`: [weighted root mean square error](https://en.wikipedia.org/wiki/Root-mean-square_deviation)
- `pearsonr`: [Pearson correlation coefficient](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient)
- `weighted_pearsonr`: [weighted Pearson correlation coefficient](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient)
- `emd`: [earth mover's distance/first Wasserstein metric](https://en.wikipedia.org/wiki/Wasserstein_metric)
- `weighted_emd`: [weighted earth mover's distance/first Wasserstein metric](https://en.wikipedia.org/wiki/Wasserstein_metric)
 

Closes #2266

Link to documentation:
- https://esmvaltool--2299.org.readthedocs.build/projects/ESMValCore/en/2299/recipe/preprocessor.html#distance-metric
- https://esmvaltool--2299.org.readthedocs.build/projects/ESMValCore/en/2299/recipe/preprocessor.html#histogram

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] ~All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful~ Codacy issue about too many arguments cannot be fixed for preprocessor function.

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by axel-lauer** [comment]
I took a first look. The example using two CMIP6 models works nicely but when I try to use ERA-Interim or ERA5 as a reference dataset I get an error message about insufficient coordinate metadata:

`ValueError: Cannot calculate distance metric between cube and reference cube: Insufficient matching coordinate metadata to resolve cubes, cannot map dimension (0,) of the RHS cube 'air_temperature' to the LHS cube 'air_temperature'.`

I have no idea what might cause this problem. This is the recipe I tried:

```
# recipe_test.yml
---
documentation:
  title: Test

  description: |
    Test new preproc.

  authors:
    - lauer_axel

  maintainer:
    - lauer_axel

  projects:
    - esmval

datasets:
  - {dataset: BCC-ESM1, project: CMIP6, exp: historical, ensemble: r1i1p1f1, grid: gn}
  - {dataset: bcc-csm1-1, version: v1, project: CMIP5, exp: historical, ensemble: r1i1p1}
#  - {dataset: ERA-Interim, project: OBS6, type: reanaly, version: '1', tier: 3, reference_for_metric: true}
  - {dataset: ERA5, project: native6, type: reanaly, version: v1, tier: 3, reference_for_metric: true}

preprocessors:
  calc_rmse:
    regrid:
      target_grid: 3x3
      scheme: linear
    distance_metric:
      metric: rmse
      coords: [latitude, longitude]
      # keep_reference_dataset: false

diagnostics:
  test:
    variables:
      tas:
        mip: Amon
        timerange: '2000/2005'
        preprocessor: calc_rmse
    scripts:
      null
```

**Comment by schlunma** [comment]
Thanks for testing Axel! I think the reason for this error are different time coordinates in the cubes. Could you try to add the preprocessor `regrid_time` to your recipe? I am not really optimistic that it will solve the problem (I don't have many good experiences with that preprocessor), but it's worth a try.

**Comment by axel-lauer** [comment]
Thanks for the idea. When using `regrid_time`, the error message changes to:

`ValueError: Cannot calculate distance metric between cube and reference cube: Coordinate 'day_of_year' has different points for the LHS cube 'air_temperature' and RHS cube 'air_temperature'.`

**Comment by schlunma** [comment]
Could you try to delete the following lines and run again?

https://github.com/ESMValGroup/ESMValCore/blob/b72da76b04c6dbf9bdb94a18bd91192e9c390321/esmvalcore/preprocessor/_time.py#L1061-L1066

**Comment by axel-lauer** [comment]
Commenting out those lines in `_time.py` results in this error message:

`ValueError: Cannot calculate distance metric between cube and reference cube: Insufficient matching coordinate metadata to resolve cubes, cannot map dimension (0,) of the RHS cube 'air_temperature' to the LHS cube 'air_temperature'.`

**Comment by schlunma** [comment]
Then I suggest to fix `regrid_time` for monthly and yearly data so that ALL cubes will end up with the exact same time coordinate. There are several open issues about this, I just commented in https://github.com/ESMValGroup/ESMValCore/issues/2106.

**Comment by axel-lauer** [comment]
Sounds like a plan!

**Comment by axel-lauer** [comment]
I found another little problem with this preprocessor: I would like to apply e.g. RMSE calculation to geographical distributions of multi-year annual means. For this, it would make sense to first use `climate_statistics` to calculate the annual means. When doing so, The variables lose their "time" dimension (the coordinate is still there but not used as a variable dimension). For this reason (I think), I get the following error message when running `distance_metric`:

```
  File "/work/bd0854/b380103/ESMValCore/esmvalcore/preprocessor/_bias.py", line 388, in _calculate_metric
    res_cube = cube.collapsed(coords, iris.analysis.MAX)
  File "/work/bd0854/b380103/mambaforge/envs/esmvaltool/lib/python3.10/site-packages/iris/cube.py", line 3888, in collapsed
    raise iris.exceptions.CoordinateCollapseError(msg)
iris.exceptions.CoordinateCollapseError: Cannot collapse a dimension which does not describe any data.
```

Any ideas what this could work? Is the `cube.collapsed` in line 388 of `_bias.py` really needed?

**Comment by schlunma** [comment]
I am not entirely sure if I understand you correctly. What are your options to `climate_statistics`? The default is to calculate the mean over the entire time period, so you will end up with just 1 value for each grid cell. In this case you cannot calculate RMSEs over time anymore (i.e., only `coords: [latitude, longitude]` would work).

If you want annual means for multiple years instead (1 value per year for each grid cell) , I think `annual_statistics` is the better choice. In this case the input for `distance_metrics` would probably be `coords: [year]`.

**Comment by axel-lauer** [comment]
Yes that's exactly what I mean:

```
    climate_statistics:
      period: full
```

This would be very useful, however, to compare annual means (e.g. for a benchmarking map plot)...

**Comment by schlunma** [comment]
Ah, so you want to get an RMSE for each grid cell that has been calculated from one time step? In that case RMSE=absolute bias so it's probably better to use the `bias` preprocessor.

I think in general it's not possible to allow calculations of distance metrics over scalar dimensions. For example, the Pearson correlation is (AFAIK) undefined for one single value (it boils down to 0/0). In addition, it's not as easy as removing the line `cube.collapsed(coords, iris.analysis.MAX)`; you would also need to make sure that the `axis` parameter (for example in `np.mean`) is handled correctly.

**Comment by axel-lauer** [comment]
I guess I have to rethink this a bit. That happens when I try to be clever...

**Comment by schlunma** [comment]
Good news: using the optimized `regrid_time` from https://github.com/ESMValGroup/ESMValCore/pull/2311 I was able to run the example recipe above that uses ERA5.

In addition, I did some tests with cubes that have different metadata. I can confirm that
- different names
- different attributes
- different cell methods
- different scalar coordinates
- different auxiliary coordinates
- different cell measures
- different auxiliary variables

are ignored. In other words, as long as the cubes have the same shape, dimensional coordinates and units, we are fine! All of that can be easily ensured with our preprocessors.

For example, I was able to calculate distance metrics for the following two cubes:

<details>
<summary>Cube 1</summary>

```
air_temperature / (K)               (time: 72; latitude: 60; longitude: 120)
    Dimension coordinates:
        time                             x             -              -
        latitude                         -             x              -
        longitude                        -             -              x
    Scalar coordinates:
        height                      2.0 m
    Cell methods:
        0                           area: time: mean (interval: 20 minutes interval: 20 minutes)
    Attributes:
        Conventions                 'CF-1.7 CMIP-6.2'
        activity_id                 'CMIP'
        branch_method               'Standard'
        branch_time_in_child        0.0
        branch_time_in_parent       2110.0
        cmor_version                '3.3.2'
        contact                     'Dr. Tongwen Wu(twwu@cma.gov.cn)'
        data_specs_version          '01.00.27'
        description                 'DECK: historical'
        experiment                  'all-forcing simulation of the recent past'
        experiment_id               'historical'
        external_variables          'areacella'
        forcing_index               1
        frequency                   'mon'
        further_info_url            'https://furtherinfo.es-doc.org/CMIP6.BCC.BCC-ESM1.historical.none.r1i1 ...'
        grid                        'T42'
        grid_label                  'gn'
        initialization_index        1
        institution                 'Beijing Climate Center, Beijing 100081, China'
        institution_id              'BCC'
        license                     'CMIP6 model data produced by BCC is licensed under a Creative Commons Attribution ...'
        mip_era                     'CMIP6'
        nominal_resolution          '250 km'
        original_name               'TREFHT'
        parent_activity_id          'CMIP'
        parent_experiment_id        'piControl'
        parent_mip_era              'CMIP6'
        parent_source_id            'BCC-ESM1'
        parent_time_units           'days since 1850-1-1 00:00:00'
        parent_variant_label        'r1i1p1f1'
        physics_index               1
        product                     'model-output'
        realization_index           1
        realm                       'atmos'
        references                  'Model described by Tongwen Wu et al. (JGR 2013; JMR 2014; submmitted to ...'
        run_variant                 'forcing: greenhouse gases,aerosol emission,solar constant,volcano mass,land ...'
        source                      'BCC-ESM 1 (2017):   aerosol: none  atmos: BCC_AGCM3_LR (T42; 128 x 64 longitude/latitude; ...'
        source_id                   'BCC-ESM1'
        source_type                 'AER AOGCM CHEM'
        sub_experiment              'none'
        sub_experiment_id           'none'
        table_id                    'Amon'
        table_info                  'Creation Date:(30 July 2018) MD5:e53ff52009d0b97d9d867dc12b6096c7'
        title                       'BCC-ESM1 output prepared for CMIP6'
        variable_id                 'tas'
        variant_label               'r1i1p1f1'
```

</details>

<details>
<summary>Cube 2</summary>

```
air_pressure / (K)                  (time: 72; latitude: 60; longitude: 120)
    Dimension coordinates:
        time                             x             -              -
        latitude                         -             x              -
        longitude                        -             -              x
    Auxiliary coordinates:
        xxxxxx                           x             -              -
    Cell measures:
        xxxxxx                           x             -              -
    Scalar coordinates:
        air_pressure                0
        height                      2.0 m
    Cell methods:
        0                           time: test
    Attributes:
        asfasfasf                   'askfljaslf'
```

</details>

**Comment by schlunma** [comment]
This is ready to review now from my side, but needs https://github.com/ESMValGroup/ESMValCore/pull/2311 to work properly for all dataset/reference combinations.

File: esmvalcore/preprocessor/_other.py [hunk 1]
```diff
@@ -85,3 +99,447 @@ def get_array_module(*args):
...
+    sorted_all_coord_dims = sorted(list(set(all_coord_dims)))
+    return tuple(sorted_all_coord_dims)
+
+
+def get_weights(
+    cube: Cube,
+    coords: Iterable[Coord] | Iterable[str],
+) -> np.ndarray | da.Array:
+    """Calculate suitable weights for given coordinates."""
+    npx = get_array_module(cube.core_data())
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_other.py:137]: this guy here `get_array_module` (and its use below, too) is causing a circular imports panic on good ole Python interpreter, via `_regrid` and `_other` (this module) - prob best bet would be to simply expand in code in your functions and not use the `get_array_module` function itself - though it's silly this way. If you asked me I'd put that function in `__init__` - it looks like an init-y function to me

**Review by valeriupredoi — CHANGES_REQUESTED** [review]

File: esmvalcore/preprocessor/_other.py (hunk 1 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_other.py]: good call! Was thinking `_shared` when I mentioned init :beers:

**Review by valeriupredoi — APPROVED** [review]
approval from me on a technical basis (Manu, pls fix merge conflicts), lemme know when Axel approves from testing/sci side, and will mergy-merge :beer:

File: doc/recipe/preprocessor.rst [hunk 2]
```diff
@@ -2649,3 +2829,74 @@ The example below shows how to set all values below zero to zero.
...
+  If ``None``, calculate the histogram over all coordinates.
+  The shape of the output cube will be `(x1, x2, ..., n_bins)`, where `xi` are
+  the dimensions of the input cube not appearing in `coords` and `n_bins` is
+  the number of bins.
+* ``bins`` (:obj:`int` or sequence of :obj:`float`, default: 10): If `bins` is
+  an :obj:`int`, it defines the number of equal-width bins in the given
+  `bin_range`.
+  If `bins` is a sequence, it defines a monotonically increasing array of bin
+  edges, including the rightmost edge, allowing for non-uniform bin widths.
+  also calculate the distance of the reference dataset with itself.
```
Comment (axel-lauer) [inline doc/recipe/preprocessor.rst:2854]:
```suggestion
```

File: doc/recipe/preprocessor.rst [hunk 3]
```diff
@@ -2649,3 +2829,74 @@ The example below shows how to set all values below zero to zero.
...
+  calculated from the input data (this only works for regular grids).
+  By default, **NO** supplementary variables will be used; they need to be
+  explicitly requested in the recipe.
+* ``normalization`` (``None``, ``'sum'``, or ``'integral'``, default:
+  ``None``): If ``None``, the result will contain the number of samples in each
+  bin.
+  If ``'integral'``, the result is the value of the probability `density`
+  function at the bin, normalized such that the integral over the range is 1.
+  If ``'sum'``, the result is the value of the probability `mass` function at
+  the bin, normalized such that the sum over the range is 1.
```
Comment (axel-lauer) [inline doc/recipe/preprocessor.rst:2886]:
```suggestion
  the bin, normalized such that the sum over the whole range is 1.
```

File: doc/recipe/preprocessor.rst [hunk 4]
```diff
@@ -2649,3 +2829,74 @@ The example below shows how to set all values below zero to zero.
...
+  the bin count (instead of 1).
+  Weights are normalized before entering the calculation if `normalization` is
+  ``'integral'`` or ``'sum'``.
+  Can be an array of the same shape as the input data, ``False`` or ``None``
+  (no weighting), or ``True``.
+  In the latter case, weighting will depend on `coords`, and the following
+  coordinates will trigger weighting: `time` (will use lengths of time
+  intervals as weights) and `latitude` (will use cell area weights).
+  Time weights are always calculated from the input data.
+  Area weights can be given as supplementary variables to the recipe
```
Comment (axel-lauer) [inline doc/recipe/preprocessor.rst:2875]:
```suggestion
  Area weights can be given as supplementary variables in the recipe
```

File: doc/recipe/preprocessor.rst [hunk 5]
```diff
@@ -2583,6 +2591,178 @@ Example:
...
+      w_i \left( x_i - \bar{x} \right) \left( r_i - \bar{r} \right)
+    }{
+      \sqrt{\sum_{i=1}^N w_i \left( x_i - \bar{x} \right)^2}
+      \sqrt{\sum_{i=1}^N w_i \left( r_i - \bar{r} \right)^2}
+    }
+
+
+  * ``'emd'``: `Unweighted Earth mover's distance`_ (EMD).
+    The EMD is also known as first Wasserstein metric `W`\ :sub:`1`, which is a
+    metric that measures distances between two probability distributions.
```
Comment (axel-lauer) [inline doc/recipe/preprocessor.rst:2666]:
```suggestion
    metric that measures distance between two probability distributions.
```

File: doc/recipe/preprocessor.rst [hunk 6]
```diff
@@ -2583,6 +2591,178 @@ Example:
...
+  .. math::
+
+    W_1 = \min_{\gamma \in \mathbb{R}^{n \times n}_{+}} \sum_{i,j}^{n}
+    \gamma_{ij} \lvert X_i - R_i \rvert \\
+    \textrm{with} ~~ \gamma 1 = p_X(X);~ \gamma^T 1 = p_R(R)
+
+  * ``'weighted_emd'``: `Weighted Earth mover's distance`_.
+    Similar to the unweighted EMD (see above), but here weights are considered
+    when calculating the probability distributions (i.e., instead of 1, each
+    element contributes a given weight in the bin count; see also `weights`
```
Comment (axel-lauer) [inline doc/recipe/preprocessor.rst:2686]:
```suggestion
    element provides a weight in the bin count; see also `weights`
```

**Review by axel-lauer — APPROVED** [review]
Looks good to me. My tests with `histogram` and `distance_metric` were successfuly and look fine. A few small wording suggestions for the documentation (please see additional comments).

**Comment by valeriupredoi** [comment]
@schlunma I fixed the merge conflict (not really something to write home about - just the weights func moved to shared but GH was confused) but also fixed one of the tests in https://github.com/ESMValGroup/ESMValCore/pull/2299/commits/2bffd035697180ca549b80f172292c53485b477d - it makes more sense as it is now, but do please have a looksee and if it's OK for you then I will merge :beer:

File: tests/unit/preprocessor/_time/test_time.py [hunk 7]
```diff
@@ -1781,7 +1781,7 @@ def test_get_time_weights_lazy():
     weights = get_time_weights(cube)
     assert isinstance(weights, da.Array)
     assert weights.shape == (2, )
-    assert weights.chunks == ((1, 1), )
+    assert weights.chunks == ((2, ), )
```
Comment (bouweandela) [inline tests/unit/preprocessor/_time/test_time.py:1784]: @valeriupredoi This looks wrong, the weights should have 2 chunks, see #2404

Comment (valeriupredoi) [reply tests/unit/preprocessor/_time/test_time.py:1784]: eh well in that case there's something :fish: y in the weights calculation?

File: esmvalcore/preprocessor/_shared.py [hunk 8]
```diff
@@ -266,3 +268,305 @@ def wrapper(data: DataType, *args: Any, **kwargs: Any) -> DataType:
...
+        raise ValueError(
+            f"Weighted statistical operations are not supported for "
+            f"{len(coord_dims):d}D time coordinates, expected 0D or 1D"
+        )
+
+    # Extract 1D time weights (= lengths of time intervals)
+    time_weights = time.lazy_bounds()[:, 1] - time.lazy_bounds()[:, 0]
+    if not cube.has_lazy_data():
+        time_weights = time_weights.compute()
+    return time_weights
```
Comment (schlunma) [inline esmvalcore/preprocessor/_shared.py:438]: I think you missed the changes of https://github.com/ESMValGroup/ESMValCore/pull/2404 here. I will re-do the merge to make sure nothing else is missing

Comment (valeriupredoi) [reply esmvalcore/preprocessor/_shared.py:438]: Hmm I merged the latest main as it was yesterday, no cherrypicking mor rebase - could git have missed it via a simple merge?

Comment (schlunma) [reply esmvalcore/preprocessor/_shared.py:438]: Bouwe changed the lines in the `_time.py`, and I moved the original function (w/o Bouwe's changes) to `_shared.py`. Thus, simply removing the function from the time module is not enough, you also need to move that to shared. Will do that now.

Comment (valeriupredoi) [reply esmvalcore/preprocessor/_shared.py]: Ouch! Indeed that then - didn't occur to me I should check the consistency of the function, sorry, bud. At least I was well verbose about what I did 😁

**Review by valeriupredoi — APPROVED** [review]

**Pull request merged**